### PR TITLE
e2e: defend on k8s version for snap-controller install

### PR DESCRIPTION
**Pull request by pkalever**
# Describe what this PR does #

Currently the scripts/install-snapshot.sh script needs to be called
depending on the Kubernetes version. It would be much easier to use the
script if it is intelligent enough to decide itself whether k8s snapshot
controller needs to be installed or not.

Fixes: #1139
Signed-off-by: Prasanna Kumar Kalever \<prasanna.kalever@redhat.com\>

### CODE REVIEW

**Comment by pkalever** [comment]
@nixpanic as mentioned this will only fix a part of issue #1139.

The second part is to detect 'what version of the snapshot components to install' right?
Where can I find the `k8s server version => snap comp version` mapping?

Thanks!

**Comment by pkalever** [comment]
I have come across this [reference](https://kubernetes-csi.github.io/docs/external-snapshotter.html#csi-external-snapshotter-sidecar) but again this doesn't have any mapping for the latest releases > v2.1.0

**Comment by pkalever** [comment]
Went through each release notes at https://github.com/kubernetes-csi/external-snapshotter/tags

If I understand it right, for our ceph-csi we need atleast k8s >= v1.14.0, 


So the map should be:
```
---------------------------------------------------------------------------
|   Kube versions                |      External snapshotter version      |
---------------------------------------------------------------------------
|      1.14.0 - v1.16.0          |                 v1.2.2                 |
---------------------------------------------------------------------------
|      1.17.0 - Latest           |              v2.1.0 or v3.0.0          |
---------------------------------------------------------------------------
```

Does it look good?

File: scripts/install-snapshot.sh
Comment (nixpanic) [inline scripts/install-snapshot.sh:98]: I think this fails when Kubernetes 2.0 gets released?

**Review by nixpanic — COMMENTED** [review]
Looks mostly good to me. Note that with the `Partially Fixes: #...` comment, the issue will be closed because GitHub detects `Fixes: #...`

File: scripts/install-snapshot.sh
Comment (pkalever) [reply scripts/install-snapshot.sh:98]: Ah! good catch, Will fix it right away.

**Comment by Madhu-1** [comment]
> Went through each release notes at https://github.com/kubernetes-csi/external-snapshotter/tags
> 
> If I understand it right, for our ceph-csi we need atleast k8s >= v1.14.0,
> 
> So the map should be:
> 
> ```
> ---------------------------------------------------------------------------
> |   Kube versions                |      External snapshotter version      |
> ---------------------------------------------------------------------------
> |      v1.14.0 - v1.16.0         |                 v1.2.2                 |
> ---------------------------------------------------------------------------
> |      v1.17.0 - Latest          |              v2.1.0 or v3.0.0          |
> ---------------------------------------------------------------------------
> ```
> 
> Does it look good?

we don't need to support snapshots if the kubernetes version is <1.17 as snapshotter beta is not backward compatible with alpha version

File: scripts/install-snapshot.sh
Comment (Madhu-1) [inline scripts/install-snapshot.sh:89]: could not get Kubernetes server version

File: scripts/install-snapshot.sh
Comment (Madhu-1) [inline scripts/install-snapshot.sh:99]: %s/kube server version/Kubernetes server version

**Comment by pkalever** [comment]
> Looks mostly good to me. Note that with the `Partially Fixes: #...` comment, the issue will be closed because GitHub detects `Fixes: #...`

Yep, I'm aware of this, my intention was to fix the other part of the issue as part of this PR.

@nixpanic @Madhu-1 @humblec can I have at least one ack on the table of mapping at https://github.com/ceph/ceph-csi/pull/1534#issuecomment-700181621

Logically:
```
if v1.14.0 <= KUBE_VERSION <= v1.16.0; then
    SNAPSHOT_VERSION=v1.2.2
elif KUBE_VERSION >= v1.17.0; then
   SNAPSHOT_VERSION=v3.0.0
fi
```

We also need to fix in the current patch, like we need to skip the snapshot controller install if k8s version < 1.14.0, rather than 1.17.0

Does this all sound good?

Thanks!

File: scripts/install-snapshot.sh
Comment (pkalever) [reply scripts/install-snapshot.sh:89]: sure

File: scripts/install-snapshot.sh
Comment (pkalever) [reply scripts/install-snapshot.sh:99]: sure

**Comment by Madhu-1** [comment]
> > Looks mostly good to me. Note that with the `Partially Fixes: #...` comment, the issue will be closed because GitHub detects `Fixes: #...`
> 
> Yep, I'm aware of this, my intention was to fix the other part of the issue as part of this PR.
> 
> @nixpanic @Madhu-1 @humblec can I have at least one ack on the table of mapping at [#1534 (comment)](https://github.com/ceph/ceph-csi/pull/1534#issuecomment-700181621)
> 
> Logically:
> 
> ```
> if v1.14.0 <= KUBE_VERSION <= v1.16.0; then
>     SNAPSHOT_VERSION=v1.2.2
> elif KUBE_VERSION >= v1.17.0; then
>    SNAPSHOT_VERSION=v3.0.0
> fi
> ```
> 
> We also need to fix in the current patch, like we need to skip the snapshot controller install if k8s version < 1.14.0, rather than 1.17.0
> 
> Does this all sound good?
> 
> Thanks!

already added here https://github.com/ceph/ceph-csi/pull/1534#issuecomment-700512294

**Comment by pkalever** [comment]
> we don't need to support snapshots if the kubernetes version is <1.17 as snapshotter beta is not backward compatible with alpha version

Thanks Madhu, this means the current patch is good enough to fix #1139

@Madhu-1 we got a new release out for External snapshotter v3.0.0 a couple of weeks ago.
I hope we are good to update the SNAPSHOT_VERSION to v3.0.0?

**Comment by Madhu-1** [comment]
> > we don't need to support snapshots if the kubernetes version is <1.17 as snapshotter beta is not backward compatible with alpha version
> 
> Thanks Madhu, this means the current patch is good enough to fix #1139
> 
> @Madhu-1 we got a new release out for External snapshotter v3.0.0 a couple of weeks ago.
> I hope we are good to update the SNAPSHOT_VERSION to v3.0.0?

Yes, @agarwal-mudit is already working on updating sidecar containers. and also can  you please update the version from v2.1.1 to 3.0.0 here https://github.com/ceph/ceph-csi/blob/d7802ebb9efbfaf26b9059f4875178494cf82861/scripts/install-snapshot.sh#L5

**Review by Madhu-1 — APPROVED** [review]

**Comment by pkalever** [comment]
CI fails with:
```
error: unable to read URL "https://raw.githubusercontent.com/kubernetes-csi/external-snapshotter/v3.0.0/config/crd/snapshot.storage.k8s.io_volumesnapshotclasses.yaml", server reported 404 Not Found, status code=404
```

Maybe we need to wait until Mudit patches land and rebase this on top of them.

**Comment by agarwal-mudit** [comment]
> CI fails with:
> 
> ```
> error: unable to read URL "https://raw.githubusercontent.com/kubernetes-csi/external-snapshotter/v3.0.0/config/crd/snapshot.storage.k8s.io_volumesnapshotclasses.yaml", server reported 404 Not Found, status code=404
> ```
> 
> Maybe we need to wait until Mudit patches land and rebase this on top of them.

Will try to raise a PR by tomorrow.

File: scripts/install-snapshot.sh
Comment (nixpanic) [inline scripts/install-snapshot.sh:5]: oh no, more versions :facepalm: These should really be part of `build.env`

**Review by nixpanic — CHANGES_REQUESTED** [review]
The failed tests complain that the v3.0.0 VolumeSnapshotClass CRD can not be found:

```
error: unable to read URL "https://raw.githubusercontent.com/kubernetes-csi/external-snapshotter/v3.0.0/config/crd/snapshot.storage.k8s.io_volumesnapshotclasses.yaml", server reported 404 Not Found, status code=404
```

Please check where the CRD needs to come from with newer versions.

File: scripts/install-snapshot.sh
Comment (pkalever) [reply scripts/install-snapshot.sh:5]: I agree with the proposal @nixpanic , will update it. Thanks!

**Comment by pkalever** [comment]
> Please check where the CRD needs to come from with newer versions.

Indeed, as dicussed in the above comments, this should be rebased on Mudit's changes, once the changes are merged. :+1:

**Comment by agarwal-mudit** [comment]
> > Please check where the CRD needs to come from with newer versions.
> 
> Indeed, as dicussed in the above comments, this should be rebased on Mudit's changes, once the changes are merged.

On a second look, is this path hard coded in the script because the crd path is changed in v3.0.0.
This is the new path: https://raw.githubusercontent.com/kubernetes-csi/external-snapshotter/v3.0.0/client/config/crd/snapshot.storage.k8s.io_volumesnapshotclasses.yaml

**Comment by Madhu-1** [comment]
@pkalever Mudit PR won't fix this one I think. we can go and make the below changes, please double-check once again

* For  2.1 below we need to use "https://raw.githubusercontent.com/kubernetes-csi/external-snapshotter/v3.0.0/config/crd
* For 3.0 onwards we need to use "https://raw.githubusercontent.com/kubernetes-csi/external-snapshotter/v3.0.0/client/config/crd

**Comment by pkalever** [comment]
@Madhu-1 you are right, starting from external snapshotter v2.0 they are at external-snapshotter/config/crd/, at v3.0 they moved to external-snapshotter/config/crd/client/.

Based on Neils's comments, I'm inclined to move the SNAPSHOT_VERSION variable to build.env.

Do we need to still support v2.1.1? Meaning do we need to defend on the versions and select the appropriate in this script? I'm not sure if we have a strong reason to support <3.0 anymore.

Thanks!

**Comment by pkalever** [comment]
let's see if the CI is happy with these changes.

**Comment by pkalever** [comment]
let's see if the CI is happy with these changes.

**Review by Madhu-1 — APPROVED** [review]

**Review by humblec — DISMISSED** [review]

File: scripts/install-snapshot.sh
Comment (Madhu-1) [inline scripts/install-snapshot.sh:12]: changing it here won't work directory is changed only for CRD, not for RBAC

**Review by Madhu-1 — CHANGES_REQUESTED** [review]

File: scripts/install-snapshot.sh
Comment (pkalever) [reply scripts/install-snapshot.sh:12]: Ah! ok, I will fix it. Thanks!

Comment (pkalever) [reply scripts/install-snapshot.sh]: done now. Thanks, Madhu.

**Comment by pkalever** [comment]
@Madhu-1 ready for merge?

**Review by Madhu-1 — APPROVED** [review]

**Comment by Madhu-1** [comment]
> @Madhu-1 ready for merge?

will wait for @nixpanic to review it.

File: build.env
Comment (humblec) [inline build.env:24]: @pkalever please mention the snapshotter release page as a source comment here, that would help

File: scripts/install-snapshot.sh
Comment (humblec) [inline scripts/install-snapshot.sh:9]: we have `3.0.1` available now @pkalever , so good to check it

File: build.env
Comment (pkalever) [reply build.env:24]: Sure Humble :+1:

File: scripts/install-snapshot.sh
Comment (nixpanic) [inline scripts/install-snapshot.sh:5]: Don't use `${BASH_SOURCE[0]}`, just use `${0}` like done in other scripts.

File: scripts/install-snapshot.sh
Comment (nixpanic) [inline scripts/install-snapshot.sh:89]: Redirecting stderr to /dev/null might make debugging more difficult in case an error occurs. Is there a good reason to not display potential errors here?

File: scripts/install-snapshot.sh
Comment (pkalever) [reply scripts/install-snapshot.sh:5]: Most of the scripts still use BASH_SOURCE,

```
[0] root 😎 04:20:27📂ceph-csi✨ cd examples/rbd/
[0] root 😎 04:20:36📂rbd✨ cd "$(dirname "${BASH_SOURCE[0]}")" >/dev/null 2>&1 && pwd
/home/pkalever/work/contributions/ceph-csi/examples/rbd
[0] root 😎 04:20:42📂rbd✨ cd "$(dirname "${0}")" >/dev/null 2>&1 && pwd
dirname: invalid option -- 'b'
Try 'dirname --help' for more information.
/home/pkalever/work/contributions/ceph-csi/examples/rbd

```

Using "${0}" seems to be erroring in the above case.

File: scripts/install-snapshot.sh
Comment (pkalever) [reply scripts/install-snapshot.sh:89]: Because we do not want the next line (condition) to return false in case of error.

File: scripts/install-snapshot.sh
Comment (pkalever) [reply scripts/install-snapshot.sh:5]: please ignore my comments above, I missed out on something and now I understand it.

**Review by humblec — DISMISSED** [review]

File: build.env
Comment (humblec) [reply build.env:24]: :+1:

File: scripts/install-snapshot.sh
Comment (nixpanic) [reply scripts/install-snapshot.sh:90]: But the output of stderr is not captured in `KUBE_VERSION` if you do not use `2>&1` or a similar redirection?

Comment (pkalever) [reply scripts/install-snapshot.sh]: fixed now. Thanks Neils.

**Review by nixpanic — APPROVED** [review]

**Review by Madhu-1 — APPROVED** [review]

**Pull request merged**
